### ISSUE TITLE
Cost history

### ISSUE THREAD

**Issue by moshloop** [flanksource/config-db#94]
# Config cost history: FOCUS-based `config_costs` table + rollup matviews

## Context

Today, cost is **four overwritten columns on `config_items`**: `cost_per_minute`, `cost_total_1d`, `cost_total_7d`, `cost_total_30d` (`duty/schema/config.hcl:279-294`). They are written by exactly one code path — the Athena/CUR `CostScraper` (`config-db/scrapers/aws/cost.go:140-200`) — which runs a four-`FULL JOIN` window query and then issues raw `UPDATE config_items SET cost_* … WHERE ? = ANY(external_id)`, bypassing `SaveResults` entirely.

This has no history, no time dimension, no currency, no service/SKU breakdown, no provenance, and no decay: once a resource stops appearing in the CUR its `cost_total_30d` stays at the last scraped value forever. `cost_per_minute` is a fiction (`last hour ÷ 60`). Unmatched spend is silently rolled onto `AWS::::Account`. And only AWS-via-Athena can produce cost at all.

**Intended outcome:** costs become an append-only history keyed by `(config_id, period_start, period_end)`, shaped on [FOCUS v1.4](https://focus.finops.org/focus-specification/v1-4/), populated by an `externalCosts` extension that *any* scraper (sql / file / exec / clickhouse / http / logs) can emit — the same seam `external_users` / `config_access` already use. The `cost_total_*` columns are dropped from `config_items` and reappear as a matview join in the `configs` view, so the existing UI and query surface keep working while gaining real history and drilldown.

---

## ⚠️ Assumption to confirm — bucketing rule

Reading "only bucket by day > 24h" as:

| Input charge period | Stored bucket |
|---|---|
| ≤ 24h (hourly, 6-hourly, daily) | **clock day** — `[date_trunc('day', start), +1 day)` |
| > 24h and ≤ 7d | **clock week** — `[date_trunc('week', start), +7 days)` |
| > 7d | **clock month** — `[date_trunc('month', start), +1 month)` |

Rows longer than 24h are **snapped, never split** into days — a monthly recurring charge stays one row. The day bucket is the only bucket sub-day input ever collapses into. Correct this in one line if it's wrong; it's isolated to `bucketFor()` (below) and its unit test.

---

### Half-open periods

FOCUS `ChargePeriodStart`/`ChargePeriodEnd` are `[start, end)` — inclusive start, **exclusive** end, always UTC. Every range query uses `>= start AND < end`; never `BETWEEN`. Bucket boundaries follow the same convention.

### Repo split

Schema is owned entirely by duty — config-db has no `.sql` files and calls `duty.migrate.RunMigrations`. So this is a **two-repo change, duty first**:

1. Land duty (schema + models + views + rbac), tag a release.
2. Bump `github.com/flanksource/duty` in config-db's `go.mod`, land the scraper/extension side.

Duty migration mechanics (`duty/migrate/migrate.go`): `functions/*.sql` → Atlas `schema/*.hcl` apply → `views/*.sql` in numeric order, with `-- dependsOn:` headers driving re-run on hash change (`duty/migrate/dependency.go`). Next free view number is **048**.

---

### `duty/schema/config_costs.hcl` (new)

Raw FOCUS line items, already merged to bucket grain. Hot dimensions are real columns; the FOCUS long tail (`Tags`, `SkuPriceDetails`, `CommitmentDiscount*`, `ContractApplied`, `Allocated*`, all `x_*`) lives in `focus jsonb`.

```hcl
table "config_costs" {
  schema = schema.public

  column "id"          { null = false type = uuid    default = sql("generate_ulid()") }
  column "config_id"   { null = true  type = uuid }   # null => unmatched, see check below
  column "scraper_id"  { null = true  type = uuid }
  column "external_id" { null = true  type = text
                         comment = "FOCUS ResourceId as scraped; retained so unmatched spend can be attached later" }

  # half-open [period_start, period_end), clock-aligned
  column "period_start" { null = false type = timestamptz }
  column "period_end"   { null = false type = timestamptz }
  column "grain"        { null = false type = text comment = "day | week | month" }

  # FOCUS dimensions kept queryable
  column "charge_category"    { null = false type = text }  # Usage|Purchase|Tax|Credit|Adjustment
  column "charge_class"       { null = true  type = text }  # 'Correction' or null
  column "service_name"       { null = true  type = text }
  column "service_category"   { null = true  type = text }
  column "sku_id"             { null = true  type = text }
  column "region_id"          { null = true  type = text }
  column "billing_account_id" { null = true  type = text }
  column "sub_account_id"     { null = true  type = text }
  column "billing_currency"   { null = false type = text }  # ISO 4217

  # FOCUS metrics — NUMERIC, never float. Wide scale: real data carries 6+ dp.
  column "billed_cost"      { null = false type = numeric(24, 10) }
  column "effective_cost"   { null = false type = numeric(24, 10) }
  column "list_cost"        { null = true  type = numeric(24, 10) }
  column "contracted_cost"  { null = true  type = numeric(24, 10) }
  column "pricing_quantity" { null = true  type = numeric(24, 10) }
  column "pricing_unit"     { null = true  type = text }

  column "focus"       { null = true  type = jsonb }
  column "fingerprint" { null = false type = text
                         comment = "deterministic hash of the dimension tuple; the merge key within a bucket" }
  column "created_at"  { null = false type = timestamptz default = sql("now()") }
  column "updated_at"  { null = false type = timestamptz default = sql("now()") }

  primary_key { columns = [column.id] }

  # idempotent merge target
  index "config_costs_merge_uniq" {
    unique  = true
    columns = [column.config_id, column.period_start, column.period_end, column.fingerprint]
  }
  index "config_costs_period_brin_idx" { type = BRIN columns = [column.period_start] }
  index "config_costs_config_period_idx" { columns = [column.config_id, column.period_start] }
  index "config_costs_unmatched_idx" {
    columns = [column.external_id]
    where   = "config_id IS NULL"
  }

  foreign_key "config_costs_config_fk" {
    columns = [column.config_id] ref_columns = [table.config_items.column.id] on_delete = CASCADE
  }
  foreign_key "config_costs_scraper_fk" {
    columns = [column.scraper_id] ref_columns = [table.config_scrapers.column.id] on_delete = SET_NULL
  }

  check "config_costs_target" { expr = "config_id IS NOT NULL OR external_id IS NOT NULL" }
  check "config_costs_period" { expr = "period_end > period_start" }
}
```

Precedent for the shape: the `check_statuses_1h` / `check_statuses_1d` rollup tables (`duty/schema/checks.hcl:347-427`) — same PK-on-bucket-start, same BRIN-on-time, same cascade FK. Ownership/provenance follows the `config_access` triad comment (`duty/schema/config_access.hcl:328-333`).

**Unmatched spend is retained, not silently rolled onto the account.** `config_id IS NULL` + `external_id` set means "we have the money, not the resource" — the compaction job re-attaches it when the config item appears, and a `config_costs_unmatched` view makes the gap visible instead of poisoning the account total.

### `duty/schema/config.hcl` — drop the four columns

Remove `cost_per_minute`, `cost_total_1d`, `cost_total_7d`, `cost_total_30d` (lines 279-294). Note `duty/schema/apply.go` filters `DropTable` but **not** `DropColumn`, so Atlas will drop these. Sequencing matters: the dependent views must be dropped in `views/000_drop.sql` before schema apply and recreated after (this is exactly what `000_drop.sql` exists for).

Leave `components.hcl:305-319` alone — component cost is out of scope and untouched.

---

### `duty/functions/cost_overlap.sql` (new)

Week/month rows straddle the 1d/7d/30d windows, so the rollup must prorate rather than include-or-exclude. One immutable helper, used everywhere:

```sql
CREATE OR REPLACE FUNCTION cost_window_overlap(
  p_start timestamptz, p_end timestamptz, w_start timestamptz, w_end timestamptz
) RETURNS numeric LANGUAGE sql IMMUTABLE PARALLEL SAFE AS $$
  SELECT GREATEST(0, EXTRACT(epoch FROM (LEAST(p_end, w_end) - GREATEST(p_start, w_start))))
       / NULLIF(EXTRACT(epoch FROM (p_end - p_start)), 0);
$$;
```

Returns the fraction of a cost row's period that falls inside the window; `cost * cost_window_overlap(...)` is the prorated contribution. A day-grain row inside the window returns 1, outside returns 0 — so the common case is exact.

### `duty/views/048_config_costs.sql` (new)

Header: `-- dependsOn: functions/cost_overlap.sql`

```sql
CREATE MATERIALIZED VIEW config_costs_rollup AS
SELECT
  config_id,
  SUM(effective_cost * cost_window_overlap(period_start, period_end, now() - interval '1 day',   now()))  AS cost_1d,
  SUM(effective_cost * cost_window_overlap(period_start, period_end, now() - interval '7 days',  now()))  AS cost_7d,
  SUM(effective_cost * cost_window_overlap(period_start, period_end, now() - interval '30 days', now()))  AS cost_30d,
  SUM(billed_cost    * cost_window_overlap(period_start, period_end, now() - interval '30 days', now()))  AS billed_30d,
  MIN(billing_currency) AS billing_currency,
  COUNT(DISTINCT billing_currency) > 1 AS mixed_currency,
  MAX(period_end) AS last_cost_at
FROM config_costs
WHERE config_id IS NOT NULL
  AND period_end >= now() - interval '30 days'
GROUP BY config_id;

CREATE UNIQUE INDEX config_costs_rollup_config_id_idx ON config_costs_rollup (config_id);

CREATE OR REPLACE FUNCTION refresh_config_costs_rollup() RETURNS void SECURITY DEFINER
LANGUAGE plpgsql AS $$ BEGIN
  REFRESH MATERIALIZED VIEW CONCURRENTLY config_costs_rollup;
END $$;
```

Mirrors `config_item_summary_3d/7d/30d` + `refresh_config_item_summary_3d()` in `duty/views/006_config_views.sql:8-115`. The unique index is what makes `CONCURRENTLY` legal.

`mixed_currency` exists because FOCUS forbids coalescing currencies — summing USD and EUR into one number is wrong, and the UI needs to know when it happened.

Plus a read view for unmatched spend:

```sql
CREATE OR REPLACE VIEW config_costs_unmatched AS
SELECT external_id, scraper_id, billing_currency,
       SUM(effective_cost) AS effective_cost, MIN(period_start) AS since, MAX(period_end) AS until
FROM config_costs WHERE config_id IS NULL
GROUP BY external_id, scraper_id, billing_currency;
```

### `duty/views/006_config_views.sql` — rewire the consumers

Header gains `-- dependsOn: views/048_config_costs.sql` so it re-runs when the matview changes.

- **`configs` view (~:135)** — drop the four passthrough columns, `LEFT JOIN config_costs_rollup`:
  ```sql
  LEFT JOIN config_costs_rollup r ON r.config_id = config_items.id
  ...
  r.cost_1d  AS cost_total_1d,
  r.cost_7d  AS cost_total_7d,
  r.cost_30d AS cost_total_30d,
  r.cost_1d / 1440.0 AS cost_per_minute,   -- honest derivation, replaces last-hour÷60
  ```
- **`config_summary` (~:357)** and **`config_class_summary` (~:418)** — `SUM(cost_total_*)` now sums the joined view columns; source these from `configs` rather than `config_items`.
- **`related_configs_recursive()` (~:779, ~:819)** and **`related_configs()` (~:855)** — the `RETURNS TABLE` signatures keep the four cost columns; the select bodies read them from `configs`. Changing a function's return type requires `DROP FUNCTION` first — add the drops to `views/000_drop.sql`.

### `duty/rbac/objects.go`

```go
"config_costs":           policy.ObjectCatalog,   // base tables block, ~:230-237
"config_costs_rollup":    policy.ObjectCatalog,   // views block, ~:16-21
"config_costs_unmatched": policy.ObjectCatalog,
"rpc/refresh_config_costs_rollup": policy.ObjectDatabaseSystem,  // ~:184-186
```

### RLS

`config_items` is RLS-enabled (`views/9998_rls_enable.sql`). The rollup matview is **not** covered by the base table's policies, so joining it into `configs` could expose cost for rows a tenant can't see. Verify against `views/035_view_rls.sql` / `035_rls_utils.sql`; if `configs` is a `security_invoker` view the join inherits correctly — if not, add an RLS policy on `config_costs` keyed the same way `config_items` is. **This is a required verification step, not an optional one.**

---

### `duty/models/config_cost.go` (new)

```go
type ConfigCost struct {
	ID          uuid.UUID  `gorm:"default:generate_ulid()"`
	ConfigID    *uuid.UUID
	ScraperID   *uuid.UUID
	ExternalID  *string
	PeriodStart time.Time
	PeriodEnd   time.Time
	Grain       string
	ChargeCategory, ChargeClass, ServiceName, ServiceCategory string
	SkuID, RegionID, BillingAccountID, SubAccountID           string
	BillingCurrency string
	BilledCost, EffectiveCost decimal.Decimal
	ListCost, ContractedCost, PricingQuantity *decimal.Decimal
	PricingUnit string
	Focus       types.JSONMap
	Fingerprint string
	CreatedAt, UpdatedAt time.Time
}
func (ConfigCost) TableName() string { return "config_costs" }

type ConfigCostRollup struct { ... }   // read-only, TableName() = "config_costs_rollup"
```

Follow `models/config_access.go` for the `types.NoOpResourceSelectable` embed + `GetFieldsMatcher/GetID/GetName/GetType` so cost rows are selectable.

### `duty/models/config.go` — remove the four fields

Drop lines 151-154 (`ConfigItem`). **Keep** `ConfigItemSummary`'s (1056-1059) — that model is backed by the `configs` view, which still exposes them. Update the `Pretty()` helper at `models/config.go:270` which renders `row["cost"]` from `CostTotal30d`.

### `duty/query/models.go` — the one real breakage

`ConfigItemQueryModel` (`:204`) targets `models.ConfigItem{}.TableName()` = `config_items` and lists the cost columns at `:207`. Dropping the columns breaks cost filtering in resource selectors.

**Fix:** point `ConfigItemQueryModel.Table` at `"configs"` (the view is a strict superset of `config_items` and is the RLS-aware read surface). `ConfigItemSummaryQueryModel` (`:236`) already targets `configs` and needs no change.

Also update:
- `duty/query/config.go:126` — `SUM(cost_total_%s)` must run against `configs`, not `config_items`. Validation at `:203` (`1d|7d|30d`) is unchanged.
- `duty/query/config_relations.go:29-32` — `RelatedConfig` keeps its four `*float64` fields, now fed by the rewired `related_configs()` function.

---

## Part 4 — config-db: the `externalCosts` extension

This follows the external-entity seam **exactly**. Reference implementations to copy, not invent: `ExternalConfigAccess` (`api/v1/interface.go:1798-1874`) and the `ConfigAccess` path through extract → fullmode → `db/permission_changes.go`.

### 4a. `api/v1` — the wire type

`api/v1/interface.go`, alongside `ExternalConfigAccess`:

```go
// +kubebuilder:object:generate=false
type ExternalCost struct {
	ConfigID         *uuid.UUID `json:"config_id,omitempty"`
	ConfigExternalID ExternalID `json:"external_config_id,omitempty"`
	ResourceID       string     `json:"resource_id,omitempty"`   // FOCUS ResourceId

	ChargePeriodStart time.Time `json:"charge_period_start"`
	ChargePeriodEnd   time.Time `json:"charge_period_end"`

	BilledCost      decimal.Decimal  `json:"billed_cost"`
	EffectiveCost   decimal.Decimal  `json:"effective_cost"`
	ListCost        *decimal.Decimal `json:"list_cost,omitempty"`
	ContractedCost  *decimal.Decimal `json:"contracted_cost,omitempty"`
	BillingCurrency string           `json:"billing_currency"`

	ChargeCategory, ChargeClass, ServiceName, ServiceCategory string
	SkuID, RegionID, BillingAccountID, SubAccountID           string
	PricingQuantity *decimal.Decimal
	PricingUnit     string

	Focus types.JSONMap `json:"-"`   // catch-all: unrecognized keys + x_*
}

func (c ExternalCost) Fingerprint() string   // sha256 over the dimension tuple
func (c ExternalCost) Validate() error
```

**`UnmarshalJSON` must accept both FOCUS PascalCase and snake_case.** A FOCUS export row (`{"ChargePeriodStart": "...", "BilledCost": 0.04}`) should paste in directly with no transform — that is the whole point of "use FOCUS as a basis". Every key that isn't a known field lands in `Focus`, which satisfies FOCUS v1.4's `CustomColumnHandling` rule that `x_*` columns must survive.

Precedent for a tolerant unmarshaller: `ExternalConfigAccess.UnmarshalJSON` (`interface.go:1855-1874`).

Add `ExternalCosts []ExternalCost \`json:"-"\`` to `ScrapeResult` (`interface.go:1281-1293` block).

### 4b. Reserved key

`scrapers/extract/schema.yaml` — new top-level reserved key `external_costs` with alias `costs` (alongside `external_users`/`users` at `:95-104`), plus a `$defs` entry. This is the *only* spec surface needed: there is no `transform.externalUsers` mapping block in this codebase — the contract is "the transform emits a config map with reserved top-level keys", gated by `spec.full: true`.

So any generic scraper works with zero new spec types:

```yaml
spec:
  full: true
  sql:
    - connection: ...
      query: SELECT ... FROM focus_export
      transform:
        expr: |
          dyn(config).map(row, {
            "external_costs": [{
              "resource_id":         row.ResourceId,
              "charge_period_start": row.ChargePeriodStart,
              "charge_period_end":   row.ChargePeriodEnd,
              "billed_cost":         row.BilledCost,
              "effective_cost":      row.EffectiveCost,
              "billing_currency":    row.BillingCurrency,
              "service_name":        row.ServiceName,
              "charge_category":     row.ChargeCategory
            }]
          }).toJSON()
```

### 4c. Extraction — `scrapers/extract/`

- `ExtractedConfig.ExternalCosts []v1.ExternalCost` (`extract.go:27-45`), plus `HasEntities()` (`:53`), `Pretty()` (`:83`), `Merge()` (`:106`).
- `fields` table entry (`extract.go:265-280`): `{&result.ExternalCosts, "external costs", []string{"external_costs", "costs"}}`. Mapping is pure `json.Marshal`/`Unmarshal` — struct tags *are* the DSL.
- `applyConfigRefDefaults` (`:670-732`) — a cost item nested under a config inherits its `uuid`/`external_id`/`type`, so the common case needs no explicit ref.
- `validateConfigRefs` (`:734-798`) — reject and `AddWarning` when: no config ref *and* no `resource_id`; `charge_period_end <= charge_period_start`; missing `billing_currency`. Increment `Summary.ExternalCosts.Skipped`.
- `ExtractionSummary` field (`:145-195`).
- `fullmode.go:58-96` — one more `entityAppender` emitting a synthetic entity-only `ScrapeResult` carrying the whole scrape's costs. Costs are **scrape-level**, same as users/groups/access.

### 4d. Bucketing — `db/external_costs.go` (new)

The merge is a **pure function**, DB-free and directly unit-testable:

```go
func bucketFor(start, end time.Time) (bStart, bEnd time.Time, grain string)
func bucketCosts(costs []v1.ExternalCost, scraperID *uuid.UUID) []models.ConfigCost
```

`bucketCosts` groups by `(config_id | external_id, bucketStart, bucketEnd, fingerprint)` and **sums** the summable metrics (`billed_cost`, `effective_cost`, `list_cost`, `contracted_cost`, `pricing_quantity`).

Never sum `ListUnitPrice` / `ContractedUnitPrice` / `PricingCurrency*UnitPrice` — FOCUS states explicitly these are row-scoped and must not be aggregated. They stay in `focus` jsonb, untouched.

### 4e. Persistence

`db/update.go`:
- `extractResult.externalCosts` (`:1625-1647`), fan-in loop (`:1669-1694`).
- New `saveExternalCosts` stage in `saveResults`, placed **after** the config-item bulk insert (`:586`) so `TempCache().FindExternalID` can resolve `ConfigExternalID` → `config_id`, and after `syncExternalEntities` (`:604`). Unresolved refs do **not** drop the row — they persist with `config_id = NULL` + `external_id`.

`db/external_costs.go` — `upsertConfigCosts`, following `upsertConfigAccess` (`db/permission_changes.go:32-307`):
- temp table `_scrape_config_costs_<scraper>` via `CREATE TEMP TABLE (LIKE config_costs INCLUDING ALL) ON COMMIT DROP` (`createTempAndInsert`, `external_entities.go:1104`)
- `duty.ApplySessionProperties` so `postgres.session.*` props apply
- ```sql
  INSERT INTO config_costs SELECT * FROM <temp>
  ON CONFLICT (config_id, period_start, period_end, fingerprint)
  DO UPDATE SET billed_cost = excluded.billed_cost,
                effective_cost = excluded.effective_cost,
                list_cost = excluded.list_cost,
                contracted_cost = excluded.contracted_cost,
                pricing_quantity = excluded.pricing_quantity,
                focus = excluded.focus,
                updated_at = now();
  ```

**Replace, not accumulate.** FOCUS providers deliver open billing periods with *Overwrite* semantics — a re-scrape of today re-states today's total. Accumulating would double-count on every scrape. Corrections arrive as their own rows (`ChargeClass = 'Correction'`, often negative) and land under a distinct fingerprint, so they add correctly.

**No stale soft-delete.** Unlike `config_access`, a period vanishing from the source is not a deletion — cost history is immutable once a billing period closes. Deletion is retention-driven only (Part 5).

### 4f. Summary plumbing

`ScrapeSummary.ExternalCosts EntitySummary[struct{}]` (`interface.go:379-383`) plus every registration point: `HasUpdates()` (`:417`), `newFormatKeys` (`:431`), `Merge` (`:621`), `summary_pretty.go` `PrettyShort()` (`:28`) and `Pretty()` (`:96`), `BuildCounts` (`:1647`), `FullScrapeResults` + `MergeScrapeResults` (`:2019`, `:2057`), `ScrapeResult.Pretty()` (`:1474`). `AsMap()` needs no change — it round-trips through JSON.

Snapshot counters: `api/v1/scrape_snapshot.go:79-128` + a `externalCostsSQL` const in `db/scrape_snapshot.go` (pattern at `:207-303`).

`db/cel_functions.go:14-28` — register `db.external_costs` / `db.external_costs_all` via the generic `queryEntities[T]` (`:96-112`) so e2e fixtures can assert on them.

### 4g. Rewrite `scrapers/aws/cost.go`

Mandatory — it writes columns that no longer exist.

- `costQueryTemplate` becomes a single per-charge-period FOCUS-shaped Athena query over the CUR: group by `(line_item_usage_start_date, line_item_usage_end_date, line_item_resource_id, line_item_product_code, product_region, line_item_usage_account_id, line_item_currency_code)` selecting `SUM(line_item_unblended_cost)` as `billed_cost` and `SUM(line_item_net_unblended_cost)` as `effective_cost`. The four `FULL JOIN` windows are deleted outright — the rollup matview replaces them.
- `Scrape` returns `v1.ScrapeResults` carrying `ExternalCosts`, instead of executing raw `UPDATE`s. All the direct-SQL code goes.
- Fix the latent bugs while there: `awsConfig.Regions[0]` panics on empty `Regions` (`:144`); `logger.Errorf(..., err)` logs a nil instead of `tx.Error` (`:174`); `strconv.ParseFloat` errors are dropped (`:111-114`) — parse into `decimal.Decimal` and fail loudly.
- Resource matching moves from `"<productCode>/<resourceId>"` + `ANY(external_id)` to the standard `ExternalID{ConfigType, ExternalID}` resolution used by every other extension.

---

## Part 5 — config-db: compaction & refresh jobs

`config-db/jobs/config_costs.go` (new), registered in `ScheduleJobs` (`jobs/jobs.go:20`) following `jobs/cleanup.go`'s `duty/job.Job` literals (`Singleton: true`, `JobHistory: true`, `Retention: job.RetentionBalanced`, counts into `ctx.History.SuccessCount`).

**`RefreshConfigCostsRollup`** — `@every 10m`, `SELECT refresh_config_costs_rollup()`. Scheduled here rather than in incident-commander's `jobs/catalog.go` because config-db owns the writes; note the precedent if it should move.

**`CompactConfigCosts`** — `@every 1h`, three passes in one transaction:

1. **Attach unmatched.** `UPDATE config_costs c SET config_id = ci.id FROM config_items ci WHERE c.config_id IS NULL AND c.external_id = ANY(ci.external_id)`. Conflicts with an existing bucket row are resolved by summing then deleting the orphan. This is what makes late-arriving resources reconcile instead of leaking into an account total.

2. **Age-based coarsening** — day → week, week → month:
   ```sql
   INSERT INTO config_costs (config_id, period_start, period_end, grain, fingerprint, ...)
   SELECT config_id, date_trunc('week', period_start), date_trunc('week', period_start) + interval '7 days',
          'week', fingerprint, SUM(billed_cost), SUM(effective_cost), ...
   FROM config_costs WHERE grain = 'day' AND period_end < now() - <threshold>
   GROUP BY config_id, 2, fingerprint, ...
   ON CONFLICT (config_id, period_start, period_end, fingerprint) DO UPDATE SET ...;
   DELETE FROM config_costs WHERE grain = 'day' AND period_end < now() - <threshold>;
   ```
   Thresholds are duty properties, **defaulting to disabled** — coarsening is lossy and should be an explicit operator choice:
   - `config.costs.compact.week.after` (suggested `90d`)
   - `config.costs.compact.month.after` (suggested `365d`)

3. **Retention.** `DELETE FROM config_costs WHERE period_end < now() - <config.costs.retention>` (default `400d`, covering 13 months so YoY comparison survives).

---

## Part 6 — Testing

Per repo convention: Ginkgo + Gomega, YAML fixtures with CEL assertions, run via `make gotest` / `make test-fast`. Unit tests colocated; DB-touching tests separate.

**Pure unit — `db/external_costs_test.go`** (table-driven, no DB). This is where the bucketing rule is pinned:

| Case | Expect |
|---|---|
| 24 hourly rows, same resource/SKU, one clock day | 1 row, `grain=day`, cost = exact sum |
| 6 hourly rows (partial day) | 1 row, `grain=day`, cost = sum of the 6 |
| row spanning `2026-08-01T00:00Z` → `2026-09-01T00:00Z` | 1 row, `grain=month`, **not split** |
| row spanning 3 days | 1 row, `grain=week` |
| row crossing midnight (23:00→01:00) | snapped to the day containing `period_start` |
| two SKUs, same resource + day | 2 rows, distinct fingerprints |
| USD + EUR, same resource + day | 2 rows — currencies never merged |
| `ChargeClass='Correction'`, negative cost | preserved, own fingerprint |
| same input applied twice | identical output (idempotence) |

**Extract unit — `scrapers/extract/testdata/unit/external_costs*.yaml`**: FOCUS PascalCase input parses; snake_case parses; `x_*` keys land in `focus`; `config_id`/`external_id` inheritance from the enclosing item; validation rejections produce warnings and bump `Skipped`.

**E2E — `scrapers/extract/testdata/e2e/external_costs_*.yaml`** via the existing harness (`scrapers/extract_e2e_test.go`), file scraper + `pre_populate.configs`:
```yaml
assertions:
  - size(db.external_costs(scraper_id)) == 1
  - db.external_costs(scraper_id)[0].grain == 'day'
  - summary.external_costs.saved == 1
```
Add `external_costs` to `e2ePrePopulate` (`extract_e2e_test.go:96-105`) and its insertion loop.

**duty — `duty/tests/config_costs_test.go`** (new, embedded Postgres): insert known `config_costs` rows and assert the rollup. Critically, **proration**: one month-grain row of \$300 spanning a 30-day month must contribute ≈\$10 to `cost_1d` and ≈\$70 to `cost_7d`, computed against an independently-derived expectation — not the query's own output.

**duty — existing tests must be migrated**, not deleted:
- `duty/tests/fixtures/dummy/config.go:134,163` set `CostTotal30d: 50` / `80` directly on `ConfigItem`; those fields are gone. Replace with `config_costs` rows that roll up to the same totals.
- `duty/tests/aggregation_test.go:72-125` (MIN/MAX/SUM over `cost_total_30d`) and `duty/tests/config_type_summary_test.go:29` must pass **unchanged** afterwards — that is the regression gate proving the view rewiring preserved behavior.

---

## Verification

```bash
# duty — schema applies cleanly, views rebuild, rollup is correct
cd ~/go/src/github.com/flanksource/duty
go test ./tests/ -run 'ConfigCost|Aggregation|ConfigTypeSummary' -v

# config-db — build (slim excludes cloud scrapers), lint, focused suites
cd ~/go/src/github.com/flanksource/config-db
make build-slim
make lint
ginkgo --tags slim -r -v --focus "external costs" ./scrapers/... ./db/...
ginkgo -r -v --focus "e2e fixture: external_costs" ./scrapers/
```

**End-to-end smoke (local, reversible — run it, don't ask):**
1. Start config-db against the embedded/local Postgres with `--db-migrations`.
2. Run a `file` scraper fixture emitting FOCUS rows for a pre-populated config item.
3. `SELECT config_id, period_start, period_end, grain, effective_cost FROM config_costs` — confirm hourly input collapsed to one day-grain row.
4. `SELECT refresh_config_costs_rollup();` then `SELECT id, cost_total_1d, cost_total_7d, cost_total_30d FROM configs WHERE id = '<config>'` — confirm the view still serves the columns the UI reads.
5. Re-run the same scrape — confirm `config_costs` row count and totals are **unchanged** (idempotence, no double-count).
6. `SELECT * FROM config_costs_unmatched` with a deliberately unknown `resource_id` — confirm the spend is visible rather than absorbed.

**AWS/Athena path** needs real credentials and a CUR table; it cannot be smoke-tested locally. Verify the rewritten `costQueryTemplate` against a recorded CUR sample in a unit test, and flag the live Athena run as the one remaining manual check.

---

## Files touched

**duty** — `schema/config_costs.hcl` (new), `schema/config.hcl` (drop 4 cols), `functions/cost_overlap.sql` (new), `views/048_config_costs.sql` (new), `views/006_config_views.sql`, `views/000_drop.sql`, `models/config_cost.go` (new), `models/config.go`, `query/models.go`, `query/config.go`, `query/config_relations.go`, `rbac/objects.go`, `tests/config_costs_test.go` (new), `tests/fixtures/dummy/config.go`

**config-db** — `api/v1/interface.go`, `api/v1/summary_pretty.go`, `api/v1/scrape_snapshot.go`, `scrapers/extract/schema.yaml`, `scrapers/extract/extract.go`, `scrapers/extract/fullmode.go`, `db/external_costs.go` (new), `db/update.go`, `db/cel_functions.go`, `db/scrape_snapshot.go`, `db/models/config_item.go` (drop 4 cols), `jobs/config_costs.go` (new), `jobs/jobs.go`, `scrapers/aws/cost.go` (rewrite), `cmd/scrapeui/{types,convert,server}.go`, plus fixtures under `scrapers/extract/testdata/{unit,e2e}/`

**Out of scope, worth flagging** — `flanksource-ui` reads `cost_total_*` from the `configs` view (`CostDetails.tsx`, `ConfigCostValue.tsx`, `ConfigListCostCell.tsx`, `ConfigSummaryList.tsx`). The view keeps serving those columns, so the UI keeps working unchanged; a cost-history drilldown on the new table is a follow-up. `components.cost_*` is untouched.
